src/Statistics.cpp

```diff
@@ -6,199 +6,139 @@
-	{
-		if (NoCaseCompare(ms_Info[i].m_Name, a_Name) == 0)
-		{
-			return ms_Info[i].m_Type;
-		}
-	}
-
-	return statInvalid;
+	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
+	return m_MainStats.at(allStatisticsNames[a_Stat]);
 }
 
 
 
 
 
-eStatistic cStatInfo::GetPrerequisite(const eStatistic a_Type)
+StatValue cStatManager::GetValue(const AString &a_Stat) const
 {
-	ASSERT((a_Type > statInvalid) && (a_Type < statCount));
+	return m_MainStats.at(a_Stat);
+}
 
-	return ms_Info[a_Type].m_Depends;
+
+
+
+
+void cStatManager::SetValue(const eStatistic &a_Stat, const StatValue a_Value)
+{
+	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
+	m_MainStats[allStatisticsNames[a_Stat]] = a_Value;
 }
 
 
 
 
 
-cStatManager::cStatManager()
+void cStatManager::SetValue(const AString &a_Stat, const StatValue a_Value)
 {
-	Reset();
+	ASSERT(std::find(allStatisticsNames.begin(), allStatisticsNames.end(), a_Stat) != allStatisticsNames.end());
+	m_MainStats[a_Stat] = a_Value;
 }
 
 
 
 
 
-StatValue cStatManager::GetValue(const eStatistic a_Stat) const
+StatValue cStatManager::AddValue(const eStatistic &a_Stat, const StatValue a_Delta)
 {
 	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
 
-	return m_MainStats[a_Stat];
+	m_MainStats[allStatisticsNames[a_Stat]] += a_Delta;
+	return m_MainStats[allStatisticsNames[a_Stat]];
 }
 
 
 
 
 
-void cStatManager::SetValue(const eStatistic a_Stat, const StatValue a_Value)
+StatValue cStatManager::AddValue(const AString &a_Stat, const StatValue a_Delta)
 {
-	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
+	ASSERT(std::find(allStatisticsNames.begin(), allStatisticsNames.end(), a_Stat) != allStatisticsNames.end());
+	m_MainStats[a_Stat] += a_Delta;
+	return m_MainStats[a_Stat];
+}
 
-	m_MainStats[a_Stat] = a_Value;
+
+
+
+
+void cStatManager::Reset(void)
+{
+	for (unsigned int i = 0; i < allStatisticsNames.size(); ++i)
+	{
+		m_MainStats[allStatisticsNames[i]] = 0;
+	}
 }
 
 
 
 
 
-StatValue cStatManager::AddValue(const eStatistic a_Stat, const StatValue a_Delta)
+const AString & cStatManager::GetName(const eStatistic a_Type)
 {
-	ASSERT((a_Stat > statInvalid) && (a_Stat < statCount));
+	return allStatisticsNames[a_Type];
+}
 
-	m_MainStats[a_Stat] += a_Delta;
 
-	return m_MainStats[a_Stat];
+
+
+
+const AString & cStatManager::GetPrerequisite(const AString & a_Type)
+{
+	return v1_8AchievementPrerequisites.at(a_Type);
 }
 
 
 
 
 
-void cStatManager::Reset(void)
+int cStatManager::GetDirtyStatistics(const std::map<AString, int> & map) const
 {
-	for (unsigned int i = 0; i < static_cast<unsigned int>(statCount); ++i)
+	int res = 0;
+	for (auto e = map.begin(); e != map.end(); e++)
 	{
-		m_MainStats[i] = 0;
+		if (GetValue(e->first))
+		{
+			res++;
+		}
 	}
+	return res;
 }
 
 
 
 
 
+int cStatManager::GetDirtyStatistics(const AStringMap & map) const
+{
+	int res = 0;
+	for (auto e = map.begin(); e != map.end(); e++)
+	{
+		if (GetValue(e->first) != 0)
```

Comment:
Iterating a map isn't the most performant. How about storing another static array of bools in cStatManager, which is set to true each time a statistic is updated?

Reply:
i implemented that map but i don't think this brough us any performance improvements - still got to check every value if it's dirty AND if it's send by this version

Reply:
Sure, we do the same amount of work in the loop, however maps don't store their data in a contiguous block like an array does, so doing a linear iteration over all the elements in a map has a higher inherent cost

Reply:
i see what you meen. as we use a C-Style Array now this should not be a concern